// @antv/g/lib/event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Util = require("@antv/util");
var Event = /** @class */ (function () {
    function Event(type, event, bubbles, cancelable) {
        this.x = 0;
        this.y = 0;
        this.clientX = 0;
        this.clientY = 0;
        this.cfg = {};
        this.bubbles = false;
        this.cancelable = false;
        this.timeStamp = new Date().getTime(); // 时间戳
        this.defaultPrevented = false;
        this.propagationStopped = false; // 阻止冒泡
        this.type = type; // 事件类型
        this.event = event;
        this.bubbles = !!bubbles;
        this.cancelable = !!cancelable;
    }
    Event.prototype.preventDefault = function () {
        this.defaultPrevented = this.cancelable && true;
    };
    Event.prototype.stopPropagation = function () {
        this.propagationStopped = true;
    };
    Event.prototype.remove = function () {
        this.cfg.removed = true;
    };
    Event.prototype.clone = function () {
        return Util.clone(this);
    };
    Event.prototype.toString = function () {
        return "[Event (type=" + this.type + ")]";
    };
    return Event;
}());
exports.default = Event;
