// @antv/g2/esm/facet/facet-sample.js
import { __extends } from "tslib";
import { Facet } from './base';
/* 一个自定义分面的示例 */
var FacetSample = /** @class */ (function (_super) {
    __extends(FacetSample, _super);
    function FacetSample(view, cfg) {
        return _super.call(this, view, cfg) || this;
    }
    FacetSample.prototype.generateFacets = function (data) {
        // your facet algorithm
        return [];
    };
    // override render
    FacetSample.prototype.render = function () {
        _super.prototype.render.call(this);
        // do what you want to add into view
    };
    // override clear
    FacetSample.prototype.clear = function () {
        _super.prototype.clear.call(this);
        // do what you want to add into view
    };
    // override destroy
    FacetSample.prototype.destroy = function () {
        _super.prototype.destroy.call(this);
        // do what you want to add into view
    };
    // implements
    FacetSample.prototype.renderTitle = function () {
        // your title component and style
    };
    // implements
    FacetSample.prototype.renderAxis = function () {
        // your axis component and style
    };
    return FacetSample;
}(Facet));
export { FacetSample };
