// @antv/g/lib/core/bbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var BBox = /** @class */ (function () {
    function BBox(x, y, width, height) {
        if (x === void 0) { x = 0; }
        if (y === void 0) { y = 0; }
        if (width === void 0) { width = 0; }
        if (height === void 0) { height = 0; }
        // range
        this.height = height;
        this.width = width;
        this.x = this.minX = x;
        this.y = this.minY = y;
        var maxX = (this.maxX = x + width);
        var maxY = (this.maxY = y + height);
        // points
        this.tl = { x: x, y: y };
        this.tr = { x: maxX, y: y };
        this.bl = { x: x, y: maxY };
        this.br = { x: maxX, y: maxY };
        // TODO
        this.bottom = 0;
        this.left = this.x;
        this.right = 0;
        this.top = this.y;
    }
    BBox.fromRange = function (minX, minY, maxX, maxY) {
        return new BBox(minX, minY, maxX - minX, maxY - minY);
    };
    BBox.prototype.equals = function (bbox) {
        return this.x === bbox.x && this.y === bbox.y && this.width === bbox.width && this.height === bbox.height;
    };
    return BBox;
}());
exports.default = BBox;
