// @antv/g2plot/esm/plots/range-column/animation.js
import * as G from '@antv/g';
import { Animate } from '@antv/g2';
import * as _ from '@antv/util';
// 记录之前的状态
var shapeCache;
function clipInFromCenterVertical(shape, animateCfg) {
    var bbox = shape.getBBox();
    var centerY = bbox.minY + bbox.height / 2;
    var cliper = new G.Rect({
        attrs: {
            x: bbox.minX,
            y: centerY,
            width: bbox.width,
            height: 0,
        },
    });
    shape.attr('clip', cliper);
    cliper.animate({
        height: bbox.height,
        y: bbox.minY,
    }, animateCfg.duration, animateCfg.easing, function () {
        shape.attr('clip', null);
    }, animateCfg.delay);
}
clipInFromCenterVertical.animationName = 'clipInFromCenterVertical';
export function setShapeCache(shapes) {
    shapeCache = shapes;
}
function updateFromCenterVertical(shape, animateCfg) {
    var fromPath = getShapeFromCache(shape).attr('path');
    var toPath = _.clone(shape.attr('path'));
    shape.attr('path', fromPath);
    shape.animate({
        path: toPath,
    }, animateCfg.duration, animateCfg.easing, animateCfg.callback, 100);
}
function getShapeFromCache(shape) {
    var id = shape.id;
    var target;
    _.each(shapeCache, function (s) {
        if (s.id === id) {
            target = s;
        }
    });
    return target;
}
updateFromCenterVertical.animationName = 'updateFromCenterVertical';
Animate.registerAnimation('appear', 'clipInFromCenterVertical', clipInFromCenterVertical);
Animate.registerAnimation('update', 'updateFromCenterVertical', updateFromCenterVertical);
