// @antv/g2/lib/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Util = require("@antv/util");
exports.Util = Util;
var G = require("@antv/g");
exports.G = G;
var Interface = require("./interface");
exports.Interface = Interface;
var plot_1 = require("./plot/plot");
exports.Plot = plot_1.default;
var view_1 = require("./plot/view");
exports.View = view_1.default;
var element_1 = require("./element");
exports.registerElement = element_1.registerElement;
exports.getElement = element_1.getElement;
exports.Element = element_1.Element;
var base_1 = require("./element/shape/base");
exports.registerShape = base_1.registerShape;
exports.registerShapeFactory = base_1.registerShapeFactory;
exports.getShapeFactory = base_1.getShapeFactory;
var facet_1 = require("./facet");
exports.registerFacet = facet_1.registerFacet;
exports.Facet = facet_1.Facet;
var interaction_1 = require("./interaction");
exports.registerInteraction = interaction_1.registerInteraction;
exports.getInteraction = interaction_1.getInteraction;
exports.Interaction = interaction_1.Interaction;
var global_1 = require("./global");
exports.Global = global_1.default;
exports.setTheme = global_1.setTheme;
exports.version = global_1.version;
var theme_1 = require("./theme");
exports.registerTheme = theme_1.registerTheme;
tslib_1.__exportStar(require("./dependents"), exports);
