// @antv/g2/lib/animate/interface.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GeomType;
(function (GeomType) {
    GeomType["line"] = "line";
    GeomType["path"] = "path";
    GeomType["area"] = "area";
    GeomType["polygon"] = "polygon";
    GeomType["edge"] = "edge";
    GeomType["interval"] = "interval";
    GeomType["point"] = "point";
    GeomType["kline"] = "kline";
    GeomType["box"] = "bbox";
    GeomType["text"] = "text";
    GeomType["heatmap"] = "heatmap";
    GeomType["label"] = "label";
    GeomType["axis-label"] = "axis-label";
    GeomType["axis-ticks"] = "axis-ticks";
    GeomType["axis-grid"] = "axis-grid";
    GeomType["axis-grid-rect"] = "axis-grid-rect";
})(GeomType = exports.GeomType || (exports.GeomType = {}));
var AnimationType;
(function (AnimationType) {
    AnimationType["appear"] = "appear";
    AnimationType["update"] = "update";
    AnimationType["enter"] = "enter";
    AnimationType["leave"] = "leave";
})(AnimationType = exports.AnimationType || (exports.AnimationType = {}));
