// @antv/scale/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var factory_1 = require("./factory");
exports.getScale = factory_1.getScale;
exports.registerScale = factory_1.registerScale;
var base_1 = require("./base");
exports.Scale = base_1.default;
var category_1 = require("./category");
exports.Category = category_1.default;
var identity_1 = require("./identity");
exports.Identity = identity_1.default;
var linear_1 = require("./linear");
exports.Linear = linear_1.default;
var linear_log_1 = require("./linear-log");
exports.Log = linear_log_1.default;
var linear_pow_1 = require("./linear-pow");
exports.Pow = linear_pow_1.default;
var time_1 = require("./time");
exports.Time = time_1.default;
var timeCat_1 = require("./timeCat");
exports.TimeCat = timeCat_1.default;
factory_1.registerScale('cat', category_1.default);
factory_1.registerScale('category', category_1.default);
factory_1.registerScale('identity', identity_1.default);
factory_1.registerScale('linear', linear_1.default);
factory_1.registerScale('log', linear_log_1.default);
factory_1.registerScale('pow', linear_pow_1.default);
factory_1.registerScale('time', time_1.default);
factory_1.registerScale('timeCat', timeCat_1.default);
