// @antv/scale/lib/linear-log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var _ = require("@antv/util");
var linear_1 = require("./linear");
var Log = /** @class */ (function (_super) {
    tslib_1.__extends(Log, _super);
    function Log() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Log.prototype.scale = function (value) {
        // 覆写主要是为了性能
        var range0 = _.head(this.range);
        var range1 = _.last(this.range);
        var percent = Math.log(value / this.min) / Math.log(this.max / this.min);
        return this._calcValue(percent, range0, range1);
    };
    Log.prototype.invert = function (scaled) {
        var range0 = _.head(this.range);
        var range1 = _.last(this.range);
        var percent = this._calcPercent(scaled, range0, range1);
        if (this.min < 0) {
            return -Math.pow(-this.max, percent) * Math.pow(-this.min, 1 - percent);
        }
        return Math.pow(this.max, percent) * Math.pow(this.min, 1 - percent);
    };
    Log.prototype._initDefaultCfg = function () {
        this.type = 'log';
        this.values = [];
        this.base = 10;
    };
    Log.prototype._init = function () {
        this._setDomain();
        // todo log breaks要优化
        this.ticks = this._setTicks();
        // 不兼容处理定义域包含0的情况
    };
    return Log;
}(linear_1.default));
exports.default = Log;
