// @antv/g2plot/esm/plots/percentage-stack-area/layer.js
import { __extends } from "tslib";
import * as _ from '@antv/util';
import { registerPlotType } from '../../base/global';
import StackArea from '../stack-area/layer';
import { transformDataPercentage } from '../../util/data';
var PercentageStackAreaLayer = /** @class */ (function (_super) {
    __extends(PercentageStackAreaLayer, _super);
    function PercentageStackAreaLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'percentageStackArea';
        return _this;
    }
    PercentageStackAreaLayer.getDefaultOptions = function () {
        return _.deepMix({}, _super.getDefaultOptions.call(this), {
            yAxis: {
                visible: true,
                label: {
                    visible: true,
                    formatter: function (v) {
                        var reg = /%/gi;
                        return v.replace(reg, '');
                    },
                },
            },
        });
    };
    PercentageStackAreaLayer.prototype.processData = function (originData) {
        var _a = this.options, xField = _a.xField, yField = _a.yField;
        return transformDataPercentage(originData, xField, [yField]);
    };
    PercentageStackAreaLayer.prototype.scale = function () {
        var metaConfig = {};
        var yField = this.options.yField;
        metaConfig[this.options.yField] = {
            tickCount: 6,
            alias: yField + " (%)",
            min: 0,
            max: 1,
            formatter: function (v) {
                var formattedValue = (v * 100).toFixed(1);
                return formattedValue + "%";
            },
        };
        this.options.meta = _.deepMix({}, metaConfig, this.options.meta);
        _super.prototype.scale.call(this);
    };
    return PercentageStackAreaLayer;
}(StackArea));
export default PercentageStackAreaLayer;
registerPlotType('percentageStackArea', PercentageStackAreaLayer);
