// @antv/g2plot/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// 通用配置
tslib_1.__exportStar(require("./interface/config"), exports);
var layer_1 = require("./base/layer");
exports.Layer = layer_1.default;
var view_layer_1 = require("./base/view-layer");
exports.ViewLayer = view_layer_1.default;
var plot_1 = require("./base/plot");
exports.Base = plot_1.default;
// 图形
tslib_1.__exportStar(require("./plots/index"), exports);
var progress_1 = require("./sparkline/progress");
exports.Progress = progress_1.default;
var ring_progress_1 = require("./sparkline/ring-progress");
exports.RingProgress = ring_progress_1.default;
var tiny_column_1 = require("./sparkline/tiny-column");
exports.TinyColumn = tiny_column_1.default;
var tiny_area_1 = require("./sparkline/tiny-area");
exports.TinyArea = tiny_area_1.default;
var tiny_line_1 = require("./sparkline/tiny-line");
exports.TinyLine = tiny_line_1.default;
// 混合图形
var overlapped_1 = require("./combo-plots/overlapped");
exports.OverlappedComboPlot = overlapped_1.default;
// 主题
var theme_1 = require("./theme");
exports.registerTheme = theme_1.registerTheme;
exports.registerGlobalTheme = theme_1.registerGlobalTheme;
var constraints_1 = require("./util/responsive/constraints");
exports.registerResponsiveConstraint = constraints_1.registerResponsiveConstraint;
var rules_1 = require("./util/responsive/rules");
exports.registerResponsiveRule = rules_1.registerResponsiveRule;
var theme_2 = require("./util/responsive/theme");
exports.registerResponsiveTheme = theme_2.registerResponsiveTheme;
exports.getResponsiveTheme = theme_2.getResponsiveTheme;
var state_manager_1 = require("./util/state-manager");
exports.StateManager = state_manager_1.default;
