// @antv/g2plot/esm/plots/range-bar/layer.js
import { __assign, __extends } from "tslib";
import * as _ from '@antv/util';
import { registerPlotType } from '../../base/global';
import BaseBarLayer from '../bar/layer';
import RangeBarLabel from './component/label';
import { setShapeCache } from './animation';
var RangeBarLayer = /** @class */ (function (_super) {
    __extends(RangeBarLayer, _super);
    function RangeBarLayer() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = 'rangeBar';
        return _this;
    }
    RangeBarLayer.getDefaultOptions = function () {
        return _.deepMix(_super.getDefaultOptions.call(this), {
            label: {
                visible: true,
                position: 'outer',
            },
            xAxis: {
                visible: true,
                autoHideLabel: false,
                autoRotateLabel: false,
                autoRotateTitle: false,
                grid: {
                    visible: true,
                },
                line: {
                    visible: false,
                },
                tickLine: {
                    visible: false,
                },
                label: {
                    visible: true,
                },
                title: {
                    visible: true,
                    offset: 12,
                },
            },
            yAxis: {
                visible: true,
                autoHideLabel: false,
                autoRotateLabel: false,
                autoRotateTitle: true,
                grid: {
                    visible: false,
                },
                line: {
                    visible: true,
                },
                tickLine: {
                    visible: true,
                },
                label: {
                    visible: true,
                },
                title: {
                    visible: false,
                    offset: 12,
                },
            },
        }, {});
    };
    RangeBarLayer.prototype.afterRender = function () {
        if (this.options.label && this.options.label.visible) {
            var label = new RangeBarLabel(__assign({ view: this.view, plot: this }, this.options.label));
            label.render();
        }
        // 为更新动画缓存shape
        var shapeCaches = [];
        var geoms = this.view.get('elements');
        _.each(geoms, function (geom) {
            var shapes = geom.getShapes();
            _.each(shapes, function (shape) {
                shapeCaches.push(shape);
            });
        });
        setShapeCache(shapeCaches);
        _super.prototype.afterRender.call(this);
    };
    RangeBarLayer.prototype.extractLabel = function () { };
    RangeBarLayer.prototype.animation = function () {
        _super.prototype.animation.call(this);
        this.bar.animate = {
            appear: {
                animation: 'clipInFromCenter',
                duration: 600,
            },
            update: {
                animation: 'updateFromCenter',
                duration: 600,
            },
        };
    };
    return RangeBarLayer;
}(BaseBarLayer));
export default RangeBarLayer;
registerPlotType('rangeBar', RangeBarLayer);
